File: csharp/chart/WhistleChat/App_Code/Response/Response.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Web;
using System.Xml;

namespace Whistle
{
    /// <summary>
    /// 抽象类,生成XML应答报文
    /// </summary>
    public abstract class Response
    {
        /// <summary>
        /// 报文类型
        /// </summary>
        public string Type
        {
            protected set;
            get;
        }

        /// <summary>
        /// 发送者
        /// </summary>
        public string FromUsername
        {
            protected set;
            get;
        }

        /// <summary>
        /// 接收者
        /// </summary>
         public string ToUsername
        {
            protected set;
             get;
        }

        /// <summary>
        /// 创建时间
        /// </summary>
         public Int64 CreateTime
        {
            protected set;
             get;
        }

        protected Request request;

        /// <summary>
        /// 构造方法
        /// </summary>
        /// <param name="request">请求内容</param>
        public Response(Request request)
        {
            this.request = request;
        }

        /// <summary>
        /// 将XML报文写入流
        /// </summary>
        /// <param name="outputStream">要写入的流</param>
        public void WriteTo(Stream outputStream)
        {
            XmlDocument doc = new XmlDocument();
     
            doc.AppendChild(doc.CreateElement("xml"));
            encode(doc);
            doc.Save(outputStream);
        }

        /// <summary>
        /// 生成XML报文
        /// </summary>
        /// <param name="doc">XML报文</param>
        protected virtual  void  encode(XmlDocument doc)
        {
            appendNode(doc["xml"], "msgid", request.MsgId);
            appendNode(doc["xml"], "fromusername", request.FromUsername);
            appendNode(doc["xml"], "tousername", request.ToUsername);
            appendNode(doc["xml"], "createtime", request.CreateTime);
        }

        /// <summary>
        /// XML报文新增string类型节点
        /// </summary>
        /// <param name="parent">父节点</param>
        /// <param name="name">新增节点名称</param>
        /// <param name="cdata">新增节点内容</param>
        /// <returns>新增节点</returns>
        protected XmlNode appendNode(XmlNode parent, string name, string cdata = null)
        {
            XmlDocument doc = parent.OwnerDocument;
            XmlNode node = parent.AppendChild(doc.CreateElement(name));

            if (cdata != null && cdata.Length > 0)
            {
                return node.AppendChild(doc.CreateCDataSection(cdata != null ? cdata : ""));
            }

            return node;
        }

        /// <summary>
        /// XML报文新增int类型节点
        /// </summary>
        /// <param name="parent">父节点</param>
        /// <param name="name">新增节点名称</param>
        /// <param name="num">新增节点内容</param>
        protected void appendNode(XmlNode parent, string name, Int64 num)
        {
            XmlDocument doc = parent.OwnerDocument;
            parent.AppendChild(doc.CreateElement(name)).AppendChild(doc.CreateCDataSection(Convert.ToString(num)));
        }
    }

    /// <summary>
    /// 文本报文生成
    /// </summary>
    public class TextResponse : Response
    {
        /// <summary>
        /// 文本内容
        /// </summary>
        public string Content
        {
             set;
             get;
        }

        /// <summary>
        /// 构造方法
        /// </summary>
        /// <param name="request">请求报文内容</param>
        public TextResponse(Request request): base(request) {}

        /// <summary>
        /// 生成XML报文
        /// </summary>
        /// <param name="doc">XML报文</param>
        protected override void encode(XmlDocument doc)
        {
            base.encode(doc);
            appendNode(doc["xml"], "msgtype", "text");
            appendNode(doc["xml"], "content", Content);
           
        }
    }

    /// <summary>
    /// 音乐报文生成
    /// </summary>
    public class MusicResponse : Response
    {
        /// <summary>
        /// 标题
        /// </summary>
        public string Title
        {
             set;
             get;
        }

        /// <summary>
        /// 描述信息
        /// </summary>
        public string Description
        {
             set;
             get;
        }

        /// <summary>
        /// 音乐链接
        /// </summary>
        public string MusicUrl
        {
             set;
             get;
        }

        /// <summary>
        /// 高品质音乐链接
        /// </summary>
        public string HQMusicUrl
        {
             set;
             get;
        }

        /// <summary>
        /// 构造方法
        /// </summary>
        /// <param name="request">请求报文</param>
        public MusicResponse(Request request) : base(request) { }

        /// <summary>
        /// 生成XML报文
        /// </summary>
        /// <param name="doc">XML报文</param>
        protected override void encode(XmlDocument doc)
        {            
             base.encode(doc);
             appendNode(doc["xml"], "msgtype", "music");

            XmlNode node = appendNode(doc["xml"], "music");
            appendNode(node, "title", Title);
            appendNode(node, "description", Description);
            appendNode(node, "musicurl", MusicUrl);
            appendNode(node, "hqmusicurl", HQMusicUrl);
           
        }
    }

    /// <summary>
    /// 图文报文
    /// </summary>
    public class Article {
        /// <summary>
        /// 标题
        /// </summary>
        public string Title {
            get;
            set;
        }

        /// <summary>
        /// 描述信息
        /// </summary>
        public string Description {
            get;
            set;
        }

        /// <summary>
        /// 图片链接
        /// </summary>
        public string PicUrl  {
            get;
            set;
        }

        /// <summary>
        /// 跳转地址
        /// </summary>
        public string Url  {
            get;
            set;
        }

        public Article() { }

        /// <summary>
        /// 构造方法
        /// </summary>
        /// <param name="title">报文Title内容</param>
        /// <param name="description">报文Description内容</param>
        /// <param name="picUrl">报文PicUrl内容</param>
        /// <param name="url">报文Url内容</param>
        public Article(string title, string description, string picUrl, string url = null) {
            Title = title;
            Description = description;
            PicUrl = picUrl;
            Url = url;
        }
    }

    /// <summary>
    /// 图文报文生成
    /// </summary>
    public class ArticleResponse : Response
    {
        /// <summary>
        /// 图文混排列表
        /// </summary>
        public List<Article> Articles
        {
             set;
             get;
        }

        /// <summary>
        /// 构造方法
        /// </summary>
        /// <param name="request">请求报文内容</param>
        public ArticleResponse(Request request) : base(request) {
            Articles = new List<Article>();
        }

        /// <summary>
        /// XML报文生成
        /// </summary>
        /// <param name="doc">XML报文</param>
        protected override void encode(XmlDocument doc)
        {
            base.encode(doc);
            appendNode(doc["xml"], "msgtype", "news");
            appendNode(doc["xml"], "articlecount", Articles.Count);

            XmlNode node = appendNode(doc["xml"], "articles");

            for (int i = 0; i < Articles.Count; ++i)
            {
                XmlNode item = appendNode(node, "item");
                appendNode(item, "title", Articles[i].Title);
                appendNode(item, "description", Articles[i].Description);
                appendNode(item, "picurl", Articles[i].PicUrl);
                appendNode(item, "url", Articles[i].Url);
            }                    
        }
    }

    /// <summary>
    /// 错误报文生成
    /// </summary>
    public class ErrorResponse : TextResponse
    {
        /// <summary>
        /// 构造方法
        /// </summary>
        /// <param name="request">请求报文内容</param>
        /// <param name="errmsg">错误消息</param>
        public ErrorResponse(Request request, string errmsg)
            : base(request)
        {           
            Content = errmsg;
        }
    }
}

File: php/chart/src/response/response.php
<?php

require_once dirname(__DIR__) . '/request/request.php';

class Response {

    protected $request;
    private $fromUserName;
    private $msgType;
    private $toUserName;
    private $createTime;

    function __construct(Request $request = null) {
        $this->request = $request;
    }

    public function getType() {
        return $this->type;
    }

    public function getFromUserName() {
        return $this->fromUserName;
    }

    public function getToUserName() {
        return $this->toUserName;
    }

    public function getCreateTime() {
        return $this->createTime;
    }

    public function setType($type) {
        $this->msgtype = $type;
    }

    public function setFromUserName($username) {
        $this->fromUserName = $username;
    }

    public function setToUserName($username) {
        $this->toUserName = $username;
    }

    public function setCreateTime($createtiime) {
        $this->createTime = $createtiime;
    }

    protected function encode(SimpleXMLElement $doc) {
        $request = $this->request;
        try {
//            $this->appendCdataNode($doc, "msgid", $this->toCdata($request->getMsgId()));
            $this->appendCdataNode($doc, "fromusername", $request->getFromUsername());
            $this->appendCdataNode($doc, "tousername", $request->getToUsername());
            $this->appendDataNode($doc, "createtime", $request->getCreateTime());
        } catch (Exception $exc) {
            echo $exc->getTraceAsString();
        }
    }

    public function addCData(SimpleXMLElement $doc, $cdata_text) {
        $node = dom_import_simplexml($doc);
        $no = $node->ownerDocument;
        $node->appendChild($no->createCDATASection($cdata_text));
    }

    protected function appendCdataNode(SimpleXMLElement $parent, $name, $cdata) {
        try {
            $tmp = $parent->addChild($name);
            $this->addCData($tmp, $cdata);
        } catch (Exception $ex) {
            echo $ex->getTraceAsString();
        }
    }

    protected function appendDataNode(SimpleXMLElement $parent, $name, $cdata) {
        try {
            $tmp = $parent->addChild($name, $cdata);
        } catch (Exception $ex) {
            echo $ex->getTraceAsString();
        }
    }

    protected function appendNode(SimpleXMLElement $parent, $name) {
        try {
            if ($parent instanceof SimpleXMLElement) {
                $node = $parent->addChild($name);
                return $node;
            }
        } catch (Exception $ex) {
            echo $ex->getTraceAsString();
        }
    }

    public function writeTo() {
        $xml = '<xml></xml>';
        $doc = new SimpleXMLElement($xml);
        $this->encode($doc);
//        $doc->asXML('D:/doc.xml');
        exit($doc->asXML());
    }

}

class TextResponse extends Response {

    private $content;

    function __construct(Request $request = null) {
        parent::__construct($request);
    }

    public function getContent() {
        return $this->content;
    }

    public function setContent($content) {
        $this->content = $content;
    }

    protected function encode(SimpleXMLElement $doc) {
        $request = $this->request;
        try {
            parent::encode($doc);
            $this->appendCdataNode($doc, "msgtype", "text");
            $this->appendCdataNode($doc, "content", $this->getContent());
        } catch (Exception $exc) {
            echo $exc->getTraceAsString();
        }
    }

}

class MusicResponse extends Response {

    private $title;
    private $description;
    private $musicurl;
    private $hqmusicurl;

    function __construct($request) {
        parent::__construct($request);
    }

    public function getTitle() {
        return $this->title;
    }

    public function setTitle($content) {
        $this->title = $content;
    }

    public function getDescription() {
        return $this->description;
    }

    public function setDescription($description) {
        $this->description = $description;
    }

    public function getMusicUrl() {
        return $this->musicurl;
    }

    public function setMusicUrl($musicurl) {
        $this->musicurl = $musicurl;
    }

    public function getHQMusicUrl() {
        return $this->hqmusicurl;
    }

    public function setHQMusicUrl($hqmusicurl) {
        $this->hqmusicurl = $hqmusicurl;
    }

    protected function encode($doc) {
        $request = $this->request;
        try {
            if ($doc instanceof SimpleXMLElement) {
                $this->appendCdataNode($doc, "msgtype", "music");
                $node = $this->appendNode($doc, "music");
                $this->appendCdataNode($node, "title", $this->getTitle());
                $this->appendCdataNode($node, "description", $this->getDescription());
                $this->appendCdataNode($node, "musicurl", $this->getMusicUrl());
                $this->appendCdataNode($node, "hqmusicurl", $this->getHQMusicUrl());
            }
        } catch (Exception $exc) {
            echo $exc->getTraceAsString();
        }
    }

}

class Article {

    private $title;
    private $description;
    private $url;
    private $picurl;

    function __construct($title, $description, $picUrl, $url = null) {
        $this->title = $title;
        $this->description = $description;
        $this->url = $url;
        $this->picurl = $picUrl;
    }

    public function getTitle() {
        return $this->title;
    }

    public function setTitle($content) {
        $this->title = $content;
    }

    public function getDescription() {
        return $this->description;
    }

    public function getUrl() {
        return $this->url;
    }

    public function setDescription($description) {
        $this->description = $description;
    }

    public function setPicUrl($picurl) {
        $this->picurl = $picurl;
    }

    public function setUrl($url) {
        return $this->url = $url;
    }

    public function getPicUrl() {
        return $this->picurl;
    }

}

class ArticleResponse extends Response {

    public $articles;

    function __construct($request) {
        parent::__construct($request);
        $this->articles = array();
    }

    public function getArticles() {
        return $this->articles;
    }

    public function setArticles($articles) {
        $this->articles = $articles;
    }

    protected function encode(SimpleXMLElement $doc) {
        $request = $this->request;
        try {
            parent::encode($doc);
            $this->appendCdataNode($doc, "msgtype", "news");
            $this->appendCdataNode($doc, "articlecount", count($this->articles));
            $node = $this->appendNode($doc, "articles");
            for ($i = 0; $i < count($this->articles); $i++) {
                $node = $this->appendNode($node, "item");
                $this->appendCdataNode($node, "title", $this->articles[$i]->getTitle());
                $this->appendCdataNode($node, "description", $this->articles[$i]->getDescription());
                $this->appendCdataNode($node, "picurl", $this->articles[$i]->getPicUrl());
                $this->appendCdataNode($node, "url", $this->articles[$i]->getUrl());
            }
        } catch (Exception $exc) {
            echo $exc->getTraceAsString();
        }
    }

}

class PushTextMessage extends TextResponse {

    private $expirationtime;

    public function __construct() {
        parent::__construct(null);
    }

    public function getExpirationTime() {
        return $this->expirationtime;
    }

    public function setExpirationTime($expirationtime) {
        return $this->expirationtime = $expirationtime;
    }

    public function encode(SimpleXMLElement $doc) {
        try {
            $this->appendCdataNode($doc, "fromusername", parent::getFromUserName());
            $this->appendCdataNode($doc, "tousername", parent::getToUsername());
            $this->appendDataNode($doc, "createtime", parent::getCreateTime());
            $this->appendCdataNode($doc, "msgtype", "text");
            $this->appendCdataNode($doc, "content", parent::getContent());
            $this->appendDataNode($doc, "expirationtime", $this->getExpirationTime());
        } catch (Exception $exc) {
            echo $exc->getTraceAsString();
        }
    }

}

class PushArticleMessage extends ArticleResponse {

    private $expirationtime;
    private $articlearray;

    public function __construct() {
        parent::__construct(null);
    }

    public function getExpirationTime() {
        return $this->expirationtime;
    }

    public function setExpirationTime($expirationtime) {
        return $this->expirationtime = $expirationtime;
    }

    public function setArticleArray($articlearray) {
        $this->articlearray = $articlearray;
    }

    public function encode(SimpleXMLElement $doc) {
        try {
            $this->appendCdataNode($doc, "fromusername", parent::getFromUserName());
            $this->appendCdataNode($doc, "tousername", parent::getToUsername());
            $this->appendDataNode($doc, "createtime", parent::getCreateTime());
            $this->appendCdataNode($doc, "msgtype", "news");
            $this->appendDataNode($doc, "expirationtime", $this->getExpirationTime());
            $this->appendCdataNode($doc, "articlecount", count($this->articlearray));
            $node = $this->appendNode($doc, "articles");
            for ($i = 0; $i < count($this->articlearray); $i++) {
                $node = $this->appendNode($node, "item");
                $arr = $this->articlearray;
                $this->appendCdataNode($node, "title", $arr[$i]->getTitle());
                $this->appendCdataNode($node, "description", $arr[$i]->getDescription());
                $this->appendCdataNode($node, "picurl", $arr[$i]->getPicUrl());
                $this->appendCdataNode($node, "url", $arr[$i]->getUrl());
            }
        } catch (Exception $exc) {
            echo $exc->getTraceAsString();
        }
    }

}
